exercises/tic-tac-toe.livemd

```diff
@@ -177,9 +189,26 @@ defmodule TicTacToe do
   iex> TicTacToe.fill([[nil, "X", nil], [nil, "X", "O"],[nil, nil, "X"]], {2, 2}, "O")
   [[nil, "X", "O"], [nil, "X", "O"],[nil, nil, "X"]]
   """
+
+  # update_at([elem], integer(), (elem -> any())) :: list() when elem: var
+
+  # find the list to update (row) within the list based on coordinates
+  # update the column (index number) of that list
+  # update the original list 
+
   def fill(board, coordinate, symbol) do
+    {x, y} = coordinate
+    board = Enum.reverse(board)
+    row = Enum.at(board, y)
+    new_row = List.update_at(row, x, &(&1 = symbol))
```

Comment:
I don't think this is working the way you want it to. I ran it locally and it doesn't produce the correct answer.
![image](https://user-images.githubusercontent.com/14877564/193926031-958ef227-ad11-4374-8d7d-0c745c14ac83.png)

## Debugging

You're really close, the only problem is that you don't rebind `board` so your function returns the unmodified reversed board.

```elixir
# BINDING board
board = Enum.reverse(board)

# IGNORED
row = Enum.at(board, y)
new_row = List.update_at(row, x, &(&1 = symbol))
List.update_at(board, y, &(&1 = new_row))

# RETURNING board
Enum.reverse(board)
```

You can resolve the issue by rebinding the `board`.

```elixir
board = Enum.reverse(board)
row = Enum.at(board, y)
new_row = List.update_at(row, x, &(&1 = symbol))
new_board = List.update_at(board, y, &(&1 = new_row))
Enum.reverse(new_board)
```

## Refactoring

Remember the `=` is the match operator, not a comparison operator.

Consider avoiding the capture operator and developing familiarity with anonymous functions as they make the return value clearer. It might be easier to debug this if it looked like the following:

```elixir
List.update_at(row, x, fn element ->
  # this is rebinding a variable, not mutating, so it simply returns `symbol`.
  element = symbol
end)
```

If you want to replace an element in a list based on it's index you can use [List.replace_at/3](https://hexdocs.pm/elixir/List.html#replace_at/3).

```elixir
board = Enum.reverse(board)
row = Enum.at(board, y)
new_row = List.replace_at(row, x, symbol)
new_board = List.replace_at(board, y, new_row)
Enum.reverse(new_board)
```

The `List.update_at/3` function is great for **retrieving** the current value and using it to **update**. We don't need to retrieve the current row since you already do that using `Enum.at/2`, which is why `List.replace_at/3` might make more sense for the above solution.

However, if you want to use the current value with `List.update_at/3` then you can update and retrieve the row in a single step.

```elixir
  def fill(board, coordinate, symbol) do
    {x, y} = coordinate
    List.update_at(board, 2 - y, fn row ->
      List.replace_at(row, x, symbol)
    end)
  end
```

Reply:
Ok, ok, so... Let's see if I have the correct understanding of rebinding... In Elixir we cannot update in the traditional sense, due to immutability. So we rebind, which means creating a new place where to store the information, hence `new_board`.

```
# IGNORED
row = Enum.at(board, y)
new_row = List.update_at(row, x, &(&1 = symbol))
List.update_at(board, y, &(&1 = new_row))
```

I do not quite understand why "#IGNORED"... Do you mean that it will get ignored because it will ultimately go to an immutable variable and not rebound to a new one?

_Consider avoiding the capture operator and developing familiarity with anonymous functions as they make the return value clearer_. - Roger that!

Having difficulty understanding the difference between update and replace.

**Update:** Returns a list with an updated value at the specified index (it retrieves and then updates by means of a function?)
**Replace:** Returns a list with a replaced value at the specified index (it doesn't retrieve? It doesn't use a function, that is what I understand).

Also... would it be possible to use pipes in the fill function? I have tried but did not manage to understand if I can rebind variables using pipes...

Reply:
> I do not quite understand why "#IGNORED"... Do you mean that it will get ignored because it will ultimately go to an immutable variable and not rebound to a new one?

By #IGNORED I mean that the code doesn't do anything, because it doesn't rebind variables and we don't have mutation it doesn't accomplish anything.

> Update: Returns a list with an updated value at the specified index (it retrieves and then updates by means of a function?)
Replace: Returns a list with a replaced value at the specified index (it doesn't retrieve? It doesn't use a function, that is what I understand).

Both accomplish similar behavior, they both update or replace a value in a list. `List.update/4` uses the previous value when updating, `List.replace_at/3` does not use the previous value and simply "replaces" it.

> Also... would it be possible to use pipes in the fill function? I have tried but did not manage to understand if I can rebind variables using pipes...

Anytime you can pass in the value as the first argument of a function, you can use the pipe operator `|>`. However whether or not you **should** depends on if it makes the code more or less clear.